### PR TITLE
Blaster refactor

### PR DESCRIPTION
**Pull request by cdsupina**
I ending up adding ManualFireComponent. I believe it is necessary because of the major differences between automatic firing, and manual firing in response to an input. I think that the way that the automatic firing cooldown needs to function vs the way the manual firing cooldown needs to function is different enough to warrant two separate components.

### CODE REVIEW

File: src/components/weapons.rs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+impl Component for AutoFireComponent {
+    type Storage = DenseVecStorage<Self>;
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct ManualFireComponent {
+    pub period: f32, // time between firing blasts
+    pub timer: f32,  // tracks time between firing blasts
+    pub ready: bool,
```
Comment (cdsupina) [inline src/components/weapons.rs:163]: ManualFireComponent has a period and timer field but also has a ready bool. This way, when the timer reaches 0.0, the ready state can be saved and used later (when the player decided to fire by pressing space).

File: src/systems/manualblaster_system.rs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+impl<'s> System<'s> for ManualBlasterSystem {
+    type SystemData = (Read<'s, Time>, WriteStorage<'s, ManualFireComponent>);
+
+    fn run(&mut self, (time, mut manual_fires): Self::SystemData) {
+        for manual_fire in (&mut manual_fires).join() {
+            if manual_fire.timer > 0.0 && !manual_fire.ready {
+                manual_fire.timer -= time.delta_seconds();
+            } else if !manual_fire.ready {
+                manual_fire.timer = manual_fire.period;
+                manual_fire.ready = true;
```
Comment (cdsupina) [inline src/systems/manualblaster_system.rs:19]: The timer is only decreased when the entity isn't ready to fire. If it is ready to fire it shouldn't be able to decrease the cooldown.

Think of it like loading a bullet in a gun. If there already is a bullet ready to be fired, you wouldn't be able to start reloading yet.

File: src/systems/spaceship.rs [hunk 3]
```diff
@@ -72,25 +74,24 @@ impl<'s> System<'s> for SpaceshipSystem {
...
-                transform,
-                hitbox.height / 2.0,
-                !spaceship.barrel_action_left && !spaceship.barrel_action_right && shoot_action,
-                time.delta_seconds(),
-            ) {
-                fire_blast(
+            if !spaceship.barrel_action_left
+                && !spaceship.barrel_action_right
+                && shoot_action
+                && manualfire.ready
```
Comment (cdsupina) [inline src/systems/spaceship.rs:85]: Checks for manualfire.ready

File: src/entities/blast.rs [hunk 4]
```diff
@@ -43,7 +32,7 @@ pub fn spawn_blasts(
         blast_transform.prepend_translation_x(blast_spacing);
     }
 }
-
+/*
 // TODO: phase out fire_blast (player/spaceship entity still uses)
```
Comment (tigleym) [inline src/entities/blast.rs:36]: Remove all this now that `fire_blast` is no longer being used

File: src/components/weapons.rs (hunk 1 above)
Comment (tigleym) [reply src/components/weapons.rs:163]: Thanks! That makes sense to me.

**Review by tigleym — APPROVED** [review]
Looks good! I think the only thing here is to remove the commented out code for `fire_blast` and also fixing those warnings.

File: src/systems/manualblaster_system.rs (hunk 2 above)
Comment (tigleym) [reply src/systems/manualblaster_system.rs:19]: Thanks for the explanation! That makes sense.

File: src/components/weapons.rs [hunk 5]
```diff
@@ -23,7 +23,8 @@ use serde::{Deserialize, Serialize};
 #[derive(Serialize, Deserialize, Debug, Clone)]
 pub struct BlasterComponent {
     pub count: usize,
-    pub allied: bool,
+    //pub allied: bool,
```
Comment (tigleym) [inline src/components/weapons.rs:26]: Let's remove this.

**Review by tigleym — APPROVED** [review]
Thanks! Using `match` for handling different blast types makes it much easier to read!

I left one comment about removing the commented out code in `weapons.rs`. Otherwise, it looks good to me!

**Pull request merged**